Make Infos and Debugs buttons appear according to present message levels (Fix Web Inspector: Infos and Debugs buttons don't appear in Console tab until new console messages are displayed)

the-chenergy opened the pull request:
#### eaa47ea7c85e3ae6230a3edb05f98821b59e6978
<pre>
Web Inspector: Infos and Debugs buttons don&apos;t appear in Console tab until new console messages are displayed
<a href="https://rdar.apple.com/122923625">rdar://122923625</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=268881">https://bugs.webkit.org/show_bug.cgi?id=268881</a>

Reviewed by Devin Rousso.

The original code hides the Infos and Debugs scope bar items until
the console receives the first message coming a non-default channel.

Make it so that the Infos and Debugs buttons&apos; visibility syncs with
with whether there are messages with those two levels respectively,
regardless of what channels the messages came from.

Call the Infos and Debugs buttons &quot;conditionally visible&quot; and make a
group for their IDs. This way the code controlling their visibility
becomes more extendable.

Also clean up some unused dead code near the primary changes.

* Source/WebInspectorUI/UserInterface/Views/LogContentView.js:
(WI.LogContentView.prototype._scopeFromMessageLevel):
  - The levels Infos and Debugs no longer depend on
    _hasNonDefaultLogChannelMessage.

(WI.LogContentView.prototype._messageAdded):
(WI.LogContentView.prototype._logCleared):
(WI.LogContentView.prototype._scopeBarSelectionDidChange):
(WI.LogContentView.prototype._showOrHideConditionallyVisibleScopeBarItemsAsNeeded):
  - Show or hide the conditionally-visible buttons as needed.

(WI.LogContentView):
  - In the constructor, hide the newly created Infos and Debugs buttons
    if they&apos;re unselected.

(WI.LogContentView.prototype._messageSourceBarSelectionDidChange):
(WI.LogContentView.prototype._scopeBarSelectionDidChange):
  - Clean up unused local variable `items`.

Canonical link: <a href="https://commits.webkit.org/275914@main">https://commits.webkit.org/275914@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ed7d185d4eec3bf41ae51a21d033aa5ed288d1d7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43029 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22055 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45431 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45654 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39154 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/25818 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19478 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35579 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43602 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19124 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37059 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16575 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/16714 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38116 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1085 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39245 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38447 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47187 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17879 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14734 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42360 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19481 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41017 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9627 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19660 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19111 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso reviewed (changes requested):
what is the rationale/motivation for this change?

In my opinion what we have now is (more) ideal because it keeps things simple for the likely lot of pages that don't use `console.info(...)` or `console.debug(...)` (and for what it's worth I would be in favor of trying to find out how often that's true just in case my intuition/experience is wrong/outdated) while still providing useful functionality in all other cases.  Perhaps you already have some data/evidence to suggest that `console.info(...)` and/or `console.debug(...)` is much more common?

If we're looking for a middle ground, perhaps we could instead have Web Inspector "remember" when `console.info(...)` and/or `console.debug(...)` is (frequently) used and eagerly show the filter items on those pages?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/LogContentView.js:
```diff
@@ -1333,6 +1323,25 @@ WI.LogContentView = class LogContentView extends WI.ContentView
             this.element.insertAdjacentElement("afterbegin", this._hiddenMessagesBannerElement);
     }
+
+    _showConditionallyVisibleScopeBarItemsIfNeeded(newMessage) {
```
Style: the `{` should be on a separate line

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/LogContentView.js:
```diff
@@ -1333,6 +1323,25 @@ WI.LogContentView = class LogContentView extends WI.ContentView
+        }
+    }
+
+    _hideConditionallyVisibleScopeBarItemsIfNeeded() {
```
ditto

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/LogContentView.js:
```diff
@@ -1333,6 +1323,25 @@ WI.LogContentView = class LogContentView extends WI.ContentView
     }
+
+    _showConditionallyVisibleScopeBarItemsIfNeeded(newMessage) {
+        for (let scope of WI.LogContentView.ConditionallyVisibleScopes) {
```
it would probably be more efficient to do this
```js
_showConditionallyVisibleScopeBarItemsIfNeeded(newMessage)
{
    let scope = this._scopeFromMessageLevel(newMessage.level);

    if (!WI.LogContentView.ConditionallyVisibleScopes.includes(scope))
        return;

    this._scopeBar.item(scope).hidden = false;
}
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/LogContentView.js:
```diff
@@ -1333,6 +1323,25 @@ WI.LogContentView = class LogContentView extends WI.ContentView
+            if (item.selected)
+                continue;
+
+            let hasMatchingMessages = this._allMessageElements().some(element => this._scopeFromMessageLevel(element.__message?.level) === scope);
```
Style: we always wrap arrow function parameters with `(` `)`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/LogContentView.js:
```diff
@@ -1333,6 +1323,25 @@ WI.LogContentView = class LogContentView extends WI.ContentView
+    }
+
+    _hideConditionallyVisibleScopeBarItemsIfNeeded() {
+        for (let scope of WI.LogContentView.ConditionallyVisibleScopes) {
```
it's probably better to iterate on `this._allMessageElements()` since that's a `querySelector`, which is more expensive and should ideally not be called more than once
```js
let messageScopes = new Set;

for (let messageElement of this._allMessageElements()) {
    let message = element.__message;
    if (!message)
        continue;

    let scope = this._scopeFromMessageLevel(message.level);
    messageScopes.add(scope);
}

for (let scope of WI.LogContentView.ConditionallyVisibleScopes) {
    let item = this._scopeBar.item(scope);
    if (item.selected)
        continue;

    if (!messageScopes.has(scope))
        item.hidden = true;
}
```

ideally we wouldn't even need to use `this._allMessageElements()`, but that would likely require an extra member variable for "what scopes have we seen so far" and some logic in `_messageAdded` (as well as all the places this is currently called) to keep it up to date

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/LogContentView.js:
```diff
@@ -1350,6 +1359,11 @@ WI.LogContentView.Scopes = {
     WebRTC: "log-webrtc",
 };
 
+WI.LogContentView.ConditionallyVisibleScopes = [
```
NIT: this might be better off as a `Set`

dcrousso reviewed (commented):
this looks much better :)

a few adjustments and i think we'd be good to go

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/LogContentView.js:
```diff
@@ -501,14 +503,12 @@ WI.LogContentView = class LogContentView extends WI.ContentView
+        let scope = this._scopeFromMessageLevel(message.level);
+        this._scopesWithMessages.add(scope);
+        if (WI.LogContentView.ConditionallyVisibleScopes.has(scope))
+            this._scopeBar.item(scope).hidden = false;
```
i think we should do this before `this._logViewController.appendConsoleMessage(message)` just in case something in there calls back out to this (e.g. `didAppendConsoleMessageView`) (i.e. set the filter before attempting to add something that uses the filter)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/LogContentView.js:
```diff
@@ -1333,6 +1329,17 @@ WI.LogContentView = class LogContentView extends WI.ContentView
+
+    _showOrHideConditionallyVisibleScopeBarItemsAsNeeded()
+    {
+        for (let scope of WI.LogContentView.ConditionallyVisibleScopes.keys()) {
```
the `.keys()` is unnecessary since it's a `Set`
```js
for (let scope of WI.LogContentView.ConditionallyVisibleScopes) {
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/LogContentView.js:
```diff
@@ -1333,6 +1329,17 @@ WI.LogContentView = class LogContentView extends WI.ContentView
+        for (let scope of WI.LogContentView.ConditionallyVisibleScopes.keys()) {
+            let item = this._scopeBar.item(scope);
+            if (item.selected)
+                continue;
```
NIT: rather than skip the below logic if `selected`, we may want to include it just in case
```js
let item = this._scopeBar.item(scope);
item.hidden = !item.selected && !this._scopesWithMessages.has(scope);
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/LogContentView.js:
```diff
@@ -501,14 +503,12 @@ WI.LogContentView = class LogContentView extends WI.ContentView
 
-        if (!this._hasNonDefaultLogChannelMessage && WI.consoleManager.customLoggingChannels.some((channel) => channel.source === message.source)) {
-            this._hasNonDefaultLogChannelMessage = true;
-            this.dispatchEventToListeners(WI.ContentView.Event.NavigationItemsDidChange);
```
ditto (to be clear we need to keep everything that isn't specifically dealing with `_scopeBar`)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/LogContentView.js:
```diff
@@ -153,7 +155,7 @@ WI.LogContentView = class LogContentView extends WI.ContentView
         ];
 
-        if (this._hasNonDefaultLogChannelMessage && this._messageSourceBar)
+        if (this._messageSourceBar)
```
i think we still need this as otherwise we're _always_ showing the channel scope bar, which we don't want to do (especially if the page doesn't use a non-default channel)

the-chenergy replied:
Whoops! I completely missed there are also message channel filters 👍

the-chenergy committed:
Web Inspector: Infos and Debugs buttons don't appear in Console tab until new console messages are displayed
rdar://122923625
https://bugs.webkit.org/show_bug.cgi?id=268881

Reviewed by Devin Rousso.

The original code hides the Infos and Debugs scope bar items until
the console receives the first message coming a non-default channel.

Make it so that the Infos and Debugs buttons' visibility syncs with
with whether there are messages with those two levels respectively,
regardless of what channels the messages came from.

Call the Infos and Debugs buttons "conditionally visible" and make a
group for their IDs. This way the code controlling their visibility
becomes more extendable.

Also clean up some unused dead code near the primary changes.

* Source/WebInspectorUI/UserInterface/Views/LogContentView.js:
(WI.LogContentView.prototype._scopeFromMessageLevel):
  - The levels Infos and Debugs no longer depend on
    _hasNonDefaultLogChannelMessage.

(WI.LogContentView.prototype._messageAdded):
(WI.LogContentView.prototype._logCleared):
(WI.LogContentView.prototype._scopeBarSelectionDidChange):
(WI.LogContentView.prototype._showOrHideConditionallyVisibleScopeBarItemsAsNeeded):
  - Show or hide the conditionally-visible buttons as needed.

(WI.LogContentView):
  - In the constructor, hide the newly created Infos and Debugs buttons
    if they're unselected.

(WI.LogContentView.prototype._messageSourceBarSelectionDidChange):
(WI.LogContentView.prototype._scopeBarSelectionDidChange):
  - Clean up unused local variable `items`.

Canonical link: https://commits.webkit.org/275914@main
